### PR TITLE
Maybe improve the parallelism of the HTTP Cache

### PR DESCRIPTION
**Pull request by gterzian**
Trying to make the cache operations more parallel across fetches. 

Assumptions:

1. Cache misses are the most likely operation, and shouldn't have to wait on accessing the cache and finding out it's a miss when another thread is actually doing a cache operation. So we introduce a `HashSet<CacheKey>` that is likely to see little contention, next to the `HashMap<CacheKey, Vec<CachedResource>>` that probably will be contented. The cost is the double storage for each key.
2. Some of the algorithms in the cache, or at least an initial setup phase, don't require locking the cache, since only the result of the algorithm affects the internal state of it(example: storing a cache entry requires checking whether it is cacheable, and then creating the entry, and finally storing it in the cache). So even if we don't go for the `HashSet<CacheKey>`, we still benefit from a slightly more granular locking inside the cache versus outside of it.
3. This are still very simple locking antics involving the entire state of the cache, so it doesn't seem to be introducing any complications. 

I can't say that I have done any measurement to validate those assumptions, any ideas on how to do so are welcome, or perhaps we can just use acceptable heuristics.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23488)

### CODE REVIEW

File: components/malloc_size_of/Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,7 @@ hashglobe = { path = "../hashglobe" }
 hyper = { version = "0.12", optional = true }
 hyper_serde = { version = "0.9", optional = true }
 keyboard-types = {version = "0.4.3", optional = true}
+parking_lot = "0.6"
```
Comment (Eijebong) [inline components/malloc_size_of/Cargo.toml]: Could this be 0.8 ? parking_lot is already duped in servo and the 0.8 version is present already (0.6 isn't though so I think your lockfile changes are invalid)

Comment (gterzian) [reply components/malloc_size_of/Cargo.toml]: Ok

**Comment by gterzian** [comment]
Ok I actually removed that extra `RwLock<HashSet<CacheKey>>`, since it creates a kind of "two-stage" update to the cache, and it could very result in a thread missing a cache update by acquiring the second lock before a thread that has already updated the first lock, but not the contents of the second. 

So it doesn't make sense because we want a "cache miss" to be fast, but not so fast that it actually results in a network request while in fact a cache hit could have happened if everything was synchronized with one lock. 

So still keeping the locking internal to the cache, since that still makes it somewhat more granular. 

Will look into the CI failures soon...

**Comment by Eijebong** [comment]
The CI failure is because the lockfile needs to be updated in the commit. You can update it by running `./mach build` locally and the committing the new changes

**Comment by gterzian** [comment]
Nah, I'll close this and will just try to keep myself busy with something more useful. 

The thing is that even a more granular lock doesn't make sense, since the absolute worst case is an extra network request. So if a thread is in the process of storing an entry in the cache, we actually want that thread to hold the lock as early as possible, in the most coarse grained manner, since just the chance of having the "waiting" thread being able to skip a networking request is worth it.

**Pull request closed without merging**